src/kestrel_datasource_stixshifter/interface.py

```diff
@@ -286,41 +300,132 @@ def query(uri, pattern, session_id, config, store=None):
+            ingestdir / f"{profile_index}_{batch_index}_{data_path_striped}.json"
+        )
+        return ingestbatchfile
+
+    return ingest_stixbundle_filepath
+
+
+async def transmission_produce(
+    transmission_queue, transmission, search_id, retrieval_batch_size, batch_index
+):
+    result_retrieval_offset = 0
+    has_remaining_results = True
+    metadata = None
+    while has_remaining_results:
+        result_batch = transmission.results(
+            search_id, result_retrieval_offset, retrieval_batch_size, metadata
+        )
+        if result_batch["success"]:
+            new_entries = result_batch["data"]
+            if new_entries:
+                result_batch["batch_index"] = batch_index
+                await transmission_queue.put(result_batch)
+                result_retrieval_offset += len(new_entries)
+                batch_index += 1
             else:
-                stixbundle = translation.translate(
-                    connector_name,
-                    "results",
-                    query_metadata,
-                    connector_results,
-                    translation_options,
-                )
+                has_remaining_results = False
+            if "metadata" in result_batch:
+                metadata = result_batch["metadata"]
+        else:
+            stix_shifter_error_msg = (
+                result_batch["error"]
+                if "error" in result_batch
+                else "details not avaliable"
+            )
+            raise DataSourceError(
+                f"STIX-shifter transmission.results() failed with message: {stix_shifter_error_msg}"
+            )
+    return batch_index
 
-                if "error" in stixbundle:
-                    raise DataSourceError(
-                        f"STIX-shifter translation results to STIX failed with message: {stixbundle['error']}"
-                    )
 
-                _logger.debug(f"dumping STIX bundles into file: {ingestfile}")
-                with ingestfile.open("w") as ingest_fp:
-                    json.dump(stixbundle, ingest_fp, indent=4)
-                bundles.append(str(ingestfile.expanduser().resolve()))
+async def translation_consume(
+    transmission_queue,
+    translation,
+    connector_name,
+    query_metadata,
+    translation_options,
+    ingest_stixbundle_filepath,
+    query_id,
+    store,
+):
+    while True:
+        # wait for an item from the producer
+        result_batch = await transmission_queue.get()
+
+        stixbundle = translation.translate(
+            connector_name,
+            "results",
+            query_metadata,
+            result_batch["data"],
+            translation_options,
+        )
+
+        if "error" in stixbundle:
+            raise DataSourceError(
+                f"STIX-shifter translation results to STIX failed with message: {stixbundle['error']}"
+            )
+
+        await asyncwrapper.SyncWrapper(store=store).cache(query_id, stixbundle)
```

Comment:
Need to check with @pcoccoli whether we have an async version of `cache()` in `firepit.aio`.

Reply:
Yes, this will work.